lessons/Assignment-3-Ruby-Part-1.md

```diff
@@ -32,76 +35,116 @@ if (x > 18) {
     console.log("x is less than or equal to 11")
 }
 ```
+
 and here is Ruby:
-```
+
+```ruby
 x = 17
 if x > 18
   puts "x is more than 18"
 elsif x > 11
-  puts "x is more than 11" 
+  puts "x is more than 11"
 else
   puts "x is less than or equal to 11"
 end
 ```
+
 As you can see, there are some syntactic differences.  For example, there is the new word "elsif", and curly braces are used less often, except in hashes.
 
+Ruby has an additional keyword, `unless`, which is effectively `if not`, and it can be paired with `else` in the same way that if does.
+
+```ruby
+x = 0
+
+unless x.positive?
+  warn "x is not a positive number"
+end
+```
+
+In Ruby, the `if`/`unless` can also come after an expression at the end of a line, but with this syntax you can't use `elsif` or `else` with it. For example:
+
+```ruby
+x = 0
+warn "x is not a positive number" unless x.positive?
+```
+
 ## Loops in Ruby
 
 There are various ways to do loops in Ruby:
-```
-3.times do
+
+```ruby
+5.times do
   puts "Hi There"
 end
 
-x = 0
-loop do
+# If you need to track which iteration of the `times` loop you're in, then you can use a "block argument."
+# You'll learn more about Ruby's "blocks" later.
+5.times do |x|
+  puts x # The value is "0-indexed", so this will print numbers 0 to 4 instead of 1 to 5
+end
+
+(1..5).each do |x|
   puts x
-  if x>7
-    break
-  end
-  x += 1
 end
 
 for i in 1..5
   puts i
 end
 
-x = 9
+x = 0
+while x < 5
+  x += 1 # Not x++ because neither `++` nor `--` work in Ruby
+  puts x
+end
 
-while x > 0 do
+x = 5
+until x < 1 # `until` is effectively `while not`
   puts x
-  x -= 1 # Not x--.  That doesn't work in Ruby
+  x -= 1
 end
 
 x = 0
+puts x while (x += 1) <= 5 # `while`/`until` can also be placed after an expression
 
-until x > 9 do
+# You can skip an iteration with the `next` keyword or exit the loop with `break`:
+
+x = 0
+while x < 5
   x += 1
-  if x == 5  # You could also have: ```if x == 5 then next``` in which case you don't need the ```end```.
-    next
-  end
+  next if x == 3 # Avoid printing the number 3
```

Comment:
Already covered the `expr if cond` syntax above, and the `then` keyword is very rare in Ruby, so I don't think we even need to mention it.